src/transaction/log_replication_atomic.cpp

```diff
@@ -381,42 +382,53 @@ namespace cublog
   void
   atomic_replicator::release_all_locks_for_ddl (cubthread::entry &thread_entry, const TRANID trid)
   {
-    assert (m_locked_classes.count (trid) > 0);
+    assert (m_locked_objects.count (trid) > 0);
 
-    for (auto classoid : m_locked_classes[trid])
+    auto [begin, end] = m_locked_objects.equal_range (trid);
+    for (auto it = begin; it != end; ++it)
       {
-	lock_unlock_object_and_cleanup (&thread_entry, &classoid, oid_Root_class_oid, SCH_M_LOCK);
+	auto log_rec = it->second;
+	lock_unlock_object_and_cleanup (&thread_entry, &log_rec.oid, &log_rec.classoid, log_rec.lock_mode);
       }
 
-    m_locked_classes.erase (trid);
+    m_locked_objects.erase (trid);
   }
 
   void
-  atomic_replicator::acquire_lock_for_ddl (cubthread::entry &thread_entry, const TRANID trid, const OID *classoid)
+  atomic_replicator::acquire_lock_for_ddl (cubthread::entry &thread_entry, const TRANID trid,
+      const LOG_REC_LOCKED_OBJECT &log_rec)
   {
-    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
+    assert (!OID_ISNULL (&log_rec.classoid) && !OID_ISTEMP (&log_rec.classoid));
+    assert (!OID_ISNULL (&log_rec.oid) && !OID_ISTEMP (&log_rec.oid));
 
     /* TODO:
      * If a PTS read transaction holds a lock for an extended period without releasing it for the same class,
      * the replicator could wait too long to acquire the lock.
      * In such cases, there might be an introduction of a mechanism to abort read transaction that holds lock.
      */
 
-    if (lock_object (&thread_entry, classoid, oid_Root_class_oid, SCH_M_LOCK, LK_UNCOND_LOCK) != LK_GRANTED)
+    if (lock_object (&thread_entry, &log_rec.oid, &log_rec.classoid, log_rec.lock_mode, LK_UNCOND_LOCK) != LK_GRANTED)
       {
 	assert_release (false);
       }
 
-    m_locked_classes[trid].emplace (*classoid);
+    m_locked_objects.emplace (trid, log_rec);
   }
 
   void
-  atomic_replicator::bookkeep_classname_for_ddl (cubthread::entry &thread_entry, const OID *classoid)
+  atomic_replicator::bookkeep_classname_for_ddl (cubthread::entry &thread_entry, const LOG_REC_LOCKED_OBJECT &log_rec)
   {
-    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
+    assert (!OID_ISNULL (&log_rec.oid) && !OID_ISTEMP (&log_rec.oid));
+
+    if (!OID_EQ (&log_rec.classoid, oid_Root_class_oid))
+      {
+	assert (OID_EQ (&log_rec.classoid, oid_Serial_class_oid));
+	return;
+      }
```

Comment:
I think it's better to do this filtering in the caller. Why should we even call this function for the db_serial class? it's not only unnecessary, but harms readability.

Reply:
OK, it will be removed